Fixes for Coverity out of bounds warnings

firewave opened the pull request.

happppp commented:
In s3c24xx.hxx, you forgot the else part.

I disagree with putting asserts whenever coverity is unsure. The point here is to look at coverity's hint and see if the code is right or not. Not suppress everything when the code is obviously fine.
For future sake when testing regularly with coverity, just keep a note locally that you have already checked those warnings.

firewave commented:
Thanks!
I am not too happy about the code where I added the asserts, but it's a pattern which is used in existing code (especially that gfx_element one) and I added those to highlight the problematic places. I think that should actually be some kind of generic finder code (lambda anyone?) and IMO it should throw an exception or fatalerror in case the lookup failed.
I only went through the "High" impact ones which should be greatly reduced after this making it easier to get an overview.

cuavas commented on src/devices/sound/ad1848.cpp:
```diff
@@ -50,7 +50,7 @@ void ad1848_device::device_start()
 void ad1848_device::device_reset()
 {
-	memset(&m_regs.idx[0], '\0', sizeof(m_regs));
+	memset(m_regs.idx, '\0', sizeof(m_regs.idx));
```
I'd prefer `std::fill(std::begin(foo), std::end(foo), 0)` for this kind of thing - it will continue to work if the type of container changes, and it will throw a fit if it's changed to a pointer rather than doing the wrong thing.

cuavas commented on src/mame/drivers/prodigy.cpp:
```diff
@@ -546,7 +546,7 @@ void prodigy_state::update_bcd()
 		case 0x04: digit_nbr = 2; break;
 		case 0x08: digit_nbr = 3; break;
-		default: logerror("Wrong BCD digit, shouldn't happen, call the maintainer!\n");
+		default: fatalerror("Wrong BCD digit, shouldn't happen, call the maintainer!\n");
```
Please don't do this - it's not bad enough to be a fatal error and will make development harder.

firewave replied:
The problem is, that digit_nbr is initialized to 4 which will cause a out-of-bounds access when it accesses ```uint8_t m_digit_cache[4]``` a few lines down (which is what Coverity complained about).

happppp replied:
You can remove default: altogether. It's impossible(both on the real chip and this emulation) for (ttl74145_data & 0x0f) to be a different value than 0,1,2,4,8.

Then, simply change:
uint8_t digit_nbr = 4;
to: uint8_t digit_nbr = 0;

Compiler and/or test tools will probably complain too if left unitialized.
